### PR TITLE
PWA-2482: Update product detail page for out of stock variations

### PR DESCRIPTION
**Pull request by cctina516**
## Description

Display out of stock variations for configurable products on product detail page. 

Preconditions

- Display Out of Stock Products = yes in Admin>Store >Configuration>Catalog>Inventory>Stock Options
- Configurable Product 1 created with all variations Out of Stock
- Configurable Product 2 created with one of the variations is Out of Stock
- Customer is on the storefront PDP 

## Related Issue

[https://jira.corp.magento.com/browse/PWA-2482](https://jira.corp.magento.com/browse/PWA-2482)

## Verification Steps

1. Display Out of Stock Products = yes in Admin>Store >Configuration>Catalog>Inventory>Stock Options
2. Create new configurable products for testing 
3. Set Configurable Product 1 created with all variations Out of Stock
4. Set Configurable Product 2 created with one/some of the variations Out of Stock

#### Test scenario(s) for direct fix/feature

1. Customer navigates to the Product 1 PDP and sees swatches related to the out of stock variations disabled and no CTA to add product to cart present
2. Customer navigate to the product 2 PDP and sees swatches active. User selects a swatch and sees related swatches updated to reflect the stock status of the variation - combination that is out of stock becomes disabled
3. User can't select a disable swatch

#### Test scenario(s) for any existing impacted features/areas

1. Customers are not able to add out of stock items to the cart
2. Customers are able to add available items to the cart

### CODE REVIEW

File: packages/venia-ui/lib/components/ProductOptions/swatchList.js [hunk 1]
```diff
@@ -6,25 +6,46 @@ import { useStyle } from '../../classify';
...
+    let isOptionOutOfStock;
     const classes = useStyle(defaultClasses, props.classes);
-
     const swatches = useMemo(
         () =>
             items.map(item => {
                 const isSelected = item.label === selectedValue.label;
 
+                if (isVariantsOutOfStock && isVariantsOutOfStock.length > 0) {
+                    const flatOosArray = isVariantsOutOfStock.flat();
```
Comment (jcalcaben) [inline packages/venia-ui/lib/components/ProductOptions/swatchList.js:25]: I would expand this out to `flatOutOfStockArray` to help reduce the cognitive load for anyone reading through the code

File: packages/venia-ui/lib/components/ProductOptions/tileList.js [hunk 2]
```diff
@@ -6,25 +6,48 @@ import { useStyle } from '../../classify';
...
+    let isOptionOutOfStock;
     const classes = useStyle(defaultClasses, props.classes);
 
     const tiles = useMemo(
         () =>
             items.map(item => {
                 const isSelected = item.label === selectedValue.label;
 
+                if (isVariantsOutOfStock && isVariantsOutOfStock.length > 0) {
+                    const flatOosArray = isVariantsOutOfStock.flat();
```
Comment (jcalcaben) [inline packages/venia-ui/lib/components/ProductOptions/tileList.js:26]: Would update this to `flatOutOfStockArray`

**Review by jcalcaben — COMMENTED** [review]
Just some initial review comments

File: packages/peregrine/lib/talons/ProductFullDetail/useProductFullDetail.js [hunk 3]
```diff
@@ -85,6 +88,111 @@ const getIsOutOfStock = (product, optionCodes, optionSelections) => {
...
+        Array.from(optionSelections.values()).filter(value => !!value).length >
+        1;
+    const outOfStockIndexes = [];
+
+    function getCombinations(arr, k, prefix = []) {
+        if (k == 0) return [prefix];
+        return arr.flatMap((v, i) =>
+            getCombinations(arr.slice(i + 1), k - 1, [...prefix, v])
+        );
+    }
```
[the remark this answers is not in the document]
Comment (cctina516) [reply packages/peregrine/lib/talons/ProductFullDetail/useProductFullDetail.js]: Thanks! It's updated.

File: packages/peregrine/lib/talons/ProductFullDetail/useProductFullDetail.js [hunk 4]
```diff
@@ -307,6 +417,21 @@ export const useProductFullDetail = props => {
         [product, optionCodes, optionSelections]
     );
 
+    const isALLOutOfStock = useMemo(() => getIsAllOutOfStock(product), [
```
[the remark this answers is not in the document]
Comment (cctina516) [reply packages/peregrine/lib/talons/ProductFullDetail/useProductFullDetail.js]: Thanks! It's updated.

File: packages/peregrine/lib/talons/ProductFullDetail/useProductFullDetail.js [hunk 5]
```diff
@@ -307,6 +417,21 @@ export const useProductFullDetail = props => {
...
+        product
+    ]);
+
+    const isVariantsOutOfStock = useMemo(
```
[the remark this answers is not in the document]
Comment (cctina516) [reply packages/peregrine/lib/talons/ProductFullDetail/useProductFullDetail.js]: Thanks! I updated the name to `outOfStockVariants`

File: packages/venia-ui/lib/components/ProductOptions/swatchList.js (hunk 1 above)
Comment (cctina516) [reply packages/venia-ui/lib/components/ProductOptions/swatchList.js]: Thanks! It's updated.

File: packages/venia-ui/lib/components/ProductOptions/tileList.js (hunk 2 above)
Comment (cctina516) [reply packages/venia-ui/lib/components/ProductOptions/tileList.js]: Thanks! It's updated.

File: packages/peregrine/lib/talons/ProductFullDetail/useProductFullDetail.js [hunk 6]
```diff
@@ -481,8 +606,12 @@ export const useProductFullDetail = props => {
             const nextOptionSelections = new Map([...optionSelections]);
             nextOptionSelections.set(optionId, selection);
             setOptionSelections(nextOptionSelections);
+            // Create a new Map to keep track of sinlge optionSelection
+            const nextSingleOptionSelection = new Map();
```
[the remark this answers is not in the document]
Comment (cctina516) [reply packages/peregrine/lib/talons/ProductFullDetail/useProductFullDetail.js:642]: The main reason I use `Map` here is to keep consistency with `optionSelections` to indicate that these are both user selections for different kind, as `optionSelections` is also using `Map`. Another advantage of `Map` is the iterability-in-order feature which may not be an obvious advantage here, but might be useful if there are new features build on this in the future.

File: packages/peregrine/lib/util/findAllMatchingVariants.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+/**
+ * TODO Document
```
Comment (jcalcaben) [inline packages/peregrine/lib/util/findAllMatchingVariants.js:2]: Document

File: packages/peregrine/lib/util/findAllMatchingVariants.js [hunk 8]
```diff
@@ -0,0 +1,34 @@
...
+    singleOptionSelection
+}) => {
+    return variants.filter(({ attributes, product }) => {
+        const customAttributes = (attributes || []).reduce(
+            (map, { code, value_index }) => new Map(map).set(code, value_index),
+            new Map()
+        );
+        for (const [id, value] of singleOptionSelection) {
+            const code = optionCodes.get(id);
+            //if (!value) return false;
```
Comment (jcalcaben) [inline packages/peregrine/lib/util/findAllMatchingVariants.js:16]: Lets remove these commented out pieces of code if we're not going to include it in the final implementation

**Review by jcalcaben — CHANGES_REQUESTED** [review]
Some additional feedback

File: packages/venia-ui/lib/components/ProductFullDetail/productFullDetail.js [hunk 9]
```diff
@@ -167,27 +171,39 @@ const ProductFullDetail = props => {
...
-            type="submit"
-        >
-            {cartCallToActionText}
-        </Button>
+        <Fragment>
+            <section className={classes.outOfStockProduct}>
+                {isEverythingOutOfStock && (
+                    <FormattedMessage
+                        data-cy="ProductFullDetail-outOfStockProduct"
+                        id={'productFullDetail.outOfStockProduct'}
```
[the remark this answers is not in the document]
Comment (cctina516) [reply packages/venia-ui/lib/components/ProductFullDetail/productFullDetail.js]: Updates -- As we decided to remove this red color error message for out of stock products on PDP, I removed this message but added error message for screen reader for accessibility. With this message removed, I also removed the entry added for this in `packages/venia-ui/i18n/en_US.json` and `packages/extensions/venia-sample-language-packs/i18n/fr_FR.json`. Please let me know if any changes should be done. Thanks!

File: packages/peregrine/lib/util/createProductVariants.js [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+
+        let variantsValueIndexes = variants.map(variant =>
+            variant.attributes.map(attribute => attribute.value_index)
+        );
+
+        let newVariantsArray = [];
+        let len = allPossibleItems.length;
+        let foundMatch;
+        let currentValueIndex = [];
+        //let attmp = [];
```
Comment (jcalcaben) [inline packages/peregrine/lib/util/createProductVariants.js:38]: don't forget to remove commented out code

File: packages/venia-ui/lib/components/ProductFullDetail/productFullDetail.js [hunk 11]
```diff
@@ -167,27 +171,36 @@ const ProductFullDetail = props => {
...
+                defaultMessage="Add to Cart"
+            />
+        ) : (
+            <FormattedMessage
+                id="productFullDetail.itemOutOfStock"
+                defaultMessage="Out of Stock"
+            />
+        );
+    // Error message for screen reader
+    const outOfStockErrorMessage = 'This item is currently out of stock';
```
Comment (jcalcaben) [inline packages/venia-ui/lib/components/ProductFullDetail/productFullDetail.js:187]: Use `formatMessage()` to make this string translate-able

File: packages/venia-ui/lib/components/ProductOptions/swatch.module.css [hunk 12]
```diff
@@ -36,3 +36,18 @@
...
+    border: 2px solid rgb(var(--venia-global-color-gray-400));
+}
+.root_outOfStock:after,
+.root_selected_outOfStock:after {
+    content: '';
+    position: absolute;
+    border-top: 2px solid rgb(var(--venia-global-color-gray-400));
+    width: 62px;
+    transform: rotate(-45deg);
+}
```
Comment (jcalcaben) [inline packages/venia-ui/lib/components/ProductOptions/swatch.module.css:53]: See if you can replace some of these values with Tailwind utilities

File: packages/venia-ui/lib/components/ProductOptions/tile.module.css [hunk 13]
```diff
@@ -25,3 +26,20 @@
...
+    composes: text-gray-600 from global;
+}
+.root_outOfStock:after,
+.root_selected_outOfStock:after {
+    content: '';
+    position: absolute;
+    border-top: 2px solid rgb(var(--venia-global-color-gray-400));
+    width: 62px;
+    transform: rotate(-45deg);
+}
```
Comment (jcalcaben) [inline packages/venia-ui/lib/components/ProductOptions/tile.module.css:45]: See if you can replace some of these values with Tailwind utilities

File: packages/peregrine/lib/util/createProductVariants.js (hunk 10 above)
Comment (cctina516) [reply packages/peregrine/lib/util/createProductVariants.js]: got it, It's updated. I will double check it going forward, thanks!

File: packages/venia-ui/lib/components/ProductFullDetail/productFullDetail.js (hunk 11 above)
Comment (cctina516) [reply packages/venia-ui/lib/components/ProductFullDetail/productFullDetail.js]: Thanks! it's updated. Also added back the entry for this in `packages/venia-ui/i18n/en_US.json` and `packages/extensions/venia-sample-language-packs/i18n/fr_FR.json`

File: packages/venia-ui/lib/components/ProductOptions/swatch.module.css (hunk 12 above)
Comment (cctina516) [reply packages/venia-ui/lib/components/ProductOptions/swatch.module.css]: Thanks! it's updated

File: packages/venia-ui/lib/components/ProductOptions/tile.module.css (hunk 13 above)
Comment (cctina516) [reply packages/venia-ui/lib/components/ProductOptions/tile.module.css]: Thanks! it's updated

File: packages/peregrine/lib/util/createProductVariants.js [hunk 14]
```diff
@@ -0,0 +1,92 @@
+/**
+ * Rebuild the array of variants with out of stock items data added.
+ * Since when admin selects in the Admin dashboard to not to display out of stock products
+ * the variants data that are needed to find disabled swatches only show the the in stock ones, missing the out of stock ones
+ * We rebuild the variants here to display all the variants and mark the stock status accordingly
+ * This returns an array of objects
+ */
+
+export const createProductVariants = product => {
+    const OUT_OF_STOCK_CODE = 'OUT_OF_STOCK';
```
[the remark this answers is not in the document]
Comment (cctina516) [reply packages/peregrine/lib/util/createProductVariants.js:10]: After discussion, we agreed to not proceed with this change.

**Review by jcalcaben — APPROVED** [review]

**Review by jcalcaben — APPROVED** [review]

**Comment by dpatil-magento** [comment]
run cypress

**Comment by dpatil-magento** [comment]
run cypress

**Comment by dpatil-magento** [comment]
run lighthouse-desktop

**Comment by dpatil-magento** [comment]
run lighthouse-mobile

**Review by dpatil-magento — APPROVED** [review]

**Pull request merged**